### ISSUE TITLE
OAuth 2.0 authorization code flow:  Access token fetch issue

### ISSUE THREAD

**Issue by harishKandikatla** [AzureAD/microsoft-authentication-library-for-js#1889]
Please follow the issue template below. Failure to do so will result in a delay in answering your question.

## Library
- [ ] `msal@1.x.x` or `@azure/msal@1.x.x`
- [ ] `@azure/msal-browser@2.x.x`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

I'm using `OAuth 2.0 authorization code flow` and not using any library

## Description
I'm using OAuth 2.0 authorization code flow in an Angular 9 application. 

I'm able to login with OAuth 2.0 authorization code flow. But **unable to get an access token**. I followed the steps from [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow)

Below is the code I used for getting token

```
    const httpOptions = {
      headers: new HttpHeaders({
        "Content-Type": "application/x-www-form-urlencoded",
      }),
    };

    const params = new HttpParams()
      .set("client_id", client_id)
      .set("redirect_uri", redirect_uri)
      .set("grant_type", grant_type)
      .set("client_secret", client_secret)
      .set("code", code);

    this.http.post<any>(url, params.toString(), httpOptions).subscribe(
      (data) => {
        console.log(data);
      },
      (err) => console.log(err)
    );
```
I'm getting **bad request** error or **CORS** error sometimes. Can you please check above implementation and correct me if I did anything wrong.

Error: Access to XMLHttpRequest at 'https://login.microsoftonline.com/{tenant}/oauth2/v2.0/token' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.

**Note** I already added this URL in the Azure portal under spa section. Also, it is working correctly in postman, I followed steps from [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow#request-an-access-token).  But not working in Angular

Questions:
> "How to get token in OAuth 2.0 authorization code flow from Angular app?"
> "Can I use msal library with OAuth 2.0 authorization code flow"?

**Comment by diljithpc** [comment]
@harishKandikatla .. are you using msal for getting authorization code? Is it possible to get the code using MSAL? even I'm trying  to do same thing. PLease share , how did you manage to given response type to code instead of id token. Thank you.

**Comment by harishKandikatla** [comment]
@diljithpc , I didn't use any library. I hope this [link](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow) helps you!

**Comment by pkanher617** [comment]
@harishKandikatla You are using a client_secret, which is a confidential client flow. This flow should not be done in a browser environment, because browsers are public clients. You will need to use authorization code flow with PKCE in order to make this work.

We suggest using the `@azure/msal-browser` [npm package](https://www.npmjs.com/package/@azure/msal-browser) if you are attempting to make this work, as this library allows you to fetch id tokens, access tokens and refresh tokens.

**Comment by pkanher617** [comment]
@diljithpc the `msal-browser` library obtains a code and trades for a token in the scope of the browser environment. If this isn't what you need (i.e. you are using this with a node backend server) we will be releasing the alpha version of the `@azure/msal-node library` shortly.

**Comment by diljithpc** [comment]
Thanks @pkanher617  and @harishKandikatla . I will need to use msal-browser for my requirement. It helped. Thank  you.

**Comment by tnorling** [comment]
Closing due to inactivity and the question was answered above. Please open a new issue if you have further questions
